### PR TITLE
Fix multi block bugs & run test-rules with single & multiple blocks

### PR DESCRIPTION
**Pull request by simonhf**
See https://github.com/VirusTotal/yara/issues/1356

### CODE REVIEW

**Comment by simonhf** [comment]
@googlebot I signed it!

File: libyara/include/yara/scanner.h [hunk 1]
```diff
@@ -92,7 +92,7 @@ YR_API int yr_scanner_scan_mem_blocks(
     YR_MEMORY_BLOCK_ITERATOR* iterator);
 
 
-YR_API int yr_scanner_scan_mem(
+YR_API int (*yr_scanner_scan_mem)(
```
Comment (plusvic) [inline libyara/include/yara/scanner.h:95]: I wouldn't do this, it changes the type of `yr_scanner_scan_mem` which is an exported function part of the API. I think you don't need to completely replace `yr_scanner_scan_mem` in tests, you can use `yr_scanner_scan_mem_blocks` providing your own iterator with custom `first` and `next` functions.

**Comment by plusvic** [comment]
Building in Mac OS is failing with:

```
duplicate symbol _yr_scanner_scan_mem in:
    .libs/exec.o
    .libs/rules.o
duplicate symbol _yr_scanner_scan_mem in:
    .libs/exec.o
    .libs/scanner.o
```

File: libyara/include/yara/scanner.h (hunk 1 above)
Comment (simonhf) [reply libyara/include/yara/scanner.h:95]: Okay. I'll find another way to override the production code biz logic with test code biz logic at test run time. And hopefully the new method will make the clang error on Mac OS go away too :-)

Comment (simonhf) [reply libyara/include/yara/scanner.h]: @plusvic Okay, so I left yr_scanner_scan_mem() intact so as not to change the API, and implemented the indirect call one level up. Finally I figured out how to appease clang running on Mac OS, and now the build and tests pass :-)

**Comment by plusvic** [comment]
@simonhf I've done some modifications to your additions in #1363, please take a look and tell me if you are ok with the changes.

**Pull request merged**